Add selection which map to to map maker

cdeil opened the pull request:
This PR adds an option to MapMaker.run to select which maps to produce, so that one can e.g. just make a counts or just exposure or just background map. The default is all maps.

This is very close now to the old map maker classes by Lea and Axel. I think that's OK; it's still the best way we know how to structure this code. (Unless someone makes a different suggestion here.) I like the two-class approach that Lea had better than the single-class approach in ImageEstimator. The big progress IMO is that we now do "cube first", and "image after" if you want, whereas the previous classes were image makers, making cubes in the background.

I didn't add an extra test, because I thought that running with a different selection like just exposure would add little extra test power. But I could add it if considered useful. I also have no opinion here on whether to keep the "make_" prefix on the methods like "make_counts", or just call those "counts", seems about equally good to me.

@registerrier - Please review.

cdeil commented:
I've changed the logic a bit in 7432cb93 . Now I delay the allocation of the zero-filled maps, moving it from init to run, to only create those if needed. So it's a bit more memory efficient (by a factor 3) if one e.g. only wants one map.

This if / else in run to make a default map is not very pretty, suggestions to organise this better welcome. On the other hand, using this class is now very simple, one always goes through `run`.

registerrier commented on gammapy/cube/make.py:
```diff
@@ -134,15 +151,42 @@ def __init__(self, obs, geom, fov_mask=None, exclusion_mask=None):
+        if selection is None:
+            selection = {'counts', 'exposure', 'background'}
+
+        if 'counts' in selection:
```
A pythonic way to avoid the series of `if` could be to do:
```
self.methods = dict(
       counts = self._make_counts,
      exposure = self._make_exposure,
      background = self._make_background
)

for map_type in selection:
       self.maps[map_type] = self.methods(map_type)()

```

registerrier commented on gammapy/cube/make.py:
```diff
@@ -95,21 +99,34 @@ def run(self, obs_list):
+
+            # Stack observation maps to total
+            if 'counts' in selection:
+                if 'counts' not in self.maps:
```
Maybe a similar solution could be applied as below in the `__init__`:
```
for map_type in selection:
      self.maps[map_type] = Map.from_geom(self.geom)
```

registerrier replied:
Of course this would work better with a test that all types passed belong to the required list

registerrier commented on gammapy/cube/make.py:
```diff
@@ -95,21 +99,34 @@ def run(self, obs_list):
+            if 'exposure' in selection:
+                if 'exposure' not in self.maps:
+                    self.maps['exposure'] = Map.from_geom(self.geom, unit="m2 s")
+                self.maps['exposure'].data[cutout_slices] += maps_obs['exposure'].quantity.to(
```
And same thing there 
```
for map_type in selection:
     self[map_type].data[cutout_slices] += map_obs[map_type].quantity.to(self.maps[map_type].unit).value
```

cdeil commented:
@registerrier - I removed the if / else, by doing method call dispatch based on selection name instead. I also added input validation for the selection option. Please review again.

cdeil commented:
> Do you think a test is needed to check failure modes for improper selections?

Would prefer to merge and move on.
